### PR TITLE
Enable kotlin support for android_library() targets

### PR DESCRIPTION
**Pull request by dsyang**
As titled. This completes #857 and enables kotlin for android!

### CODE REVIEW

File: src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java [hunk 1]
```diff
@@ -16,22 +16,27 @@
...
 import com.facebook.buck.util.HumanReadableException;
 
 public class DefaultAndroidLibraryCompilerFactory implements AndroidLibraryCompilerFactory {
 
   private final ScalaBuckConfig scalaConfig;
+  private final KotlinBuckConfig kotlinBuckConfig;
 
-  public DefaultAndroidLibraryCompilerFactory(ScalaBuckConfig scalaConfig) {
+  public DefaultAndroidLibraryCompilerFactory(
+      ScalaBuckConfig scalaConfig, KotlinBuckConfig kotlinBuckConfig) {
```
Comment (dsyang) [inline src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java]: Will break this out into multiple lines

File: src/com/facebook/buck/jvm/java/BaseCompileToJarStepFactory.java [hunk 2]
```diff
@@ -40,6 +41,14 @@
  */
 public abstract class BaseCompileToJarStepFactory implements CompileToJarStepFactory {
 
+  public static final Function<BuildContext, Iterable<Path>> EMPTY_EXTRA_CLASSPATH =
```
Comment (dsyang) [inline src/com/facebook/buck/jvm/java/BaseCompileToJarStepFactory.java]: will update this to java 8 syntax

File: src/com/facebook/buck/jvm/scala/ScalacToJarStepFactory.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-/*
+<<<</*
```
Comment (dsyang) [inline src/com/facebook/buck/jvm/scala/ScalacToJarStepFactory.java]: oops. will remove

File: test/com/facebook/buck/android/AndroidLibraryIntegrationTest.java [hunk 4]
```diff
@@ -53,6 +54,26 @@ public void testAndroidLibraryDoesNotUseTransitiveResources() throws IOException
     assertTrue(result.getStderr().contains("package R does not exist"));
   }
 
+  @Test()
```
Comment (dsyang) [inline test/com/facebook/buck/android/AndroidLibraryIntegrationTest.java]: will remove extra parens

File: test/com/facebook/buck/android/testdata/android_project/kotlin/BUCK.fixture [hunk 5]
```diff
@@ -0,0 +1,5 @@
+prebuilt_jar(
+  name='runtime',
+  binary_jar='kotlin-runtime.jar',
+  visibility= [ 'PUBLIC' ],
+)
```
Comment (dsyang) [inline test/com/facebook/buck/android/testdata/android_project/kotlin/BUCK.fixture]: will add newline

File: test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/Activity.kt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package com.sample.lib
+
+object Activity {
+    var myString = "__FIND_ME_HERE__content=1__GOODBYE__"
+}
```
Comment (dsyang) [inline test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/Activity.kt]: same here

File: test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/Sample2.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import android.os.Bundle
+
+class Sample2 : Activity() {
+
+    override fun onCreate(savedInstanceState: Bundle) {
+        super.onCreate(savedInstanceState)
+
+        setTitle(R.string.title)
+    }
+}
```
Comment (dsyang) [inline test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/Sample2.kt]: add newline

File: test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/BUCK.fixture [hunk 8]
```diff
@@ -0,0 +1,35 @@
+android_library(
+  name = 'lib',
+  srcs = glob(['*.kt']),
+  language = 'KOTLIN',
```
Comment (dsyang) [inline test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/BUCK.fixture]: We should test this but my feeling is java_library (and by association android_library before #913) expected all of it's sources to be java files. I've made the languages param optional in the latest update to this PR so we can see what that would be like. After testing the old behavior of java/scala we should decide what to do.

Comment (bolinfest) [inline test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/BUCK.fixture]: The only thing `java_library()` was allowed to receive in its `srcs` in the past were files that ended in `.java` or `.src.zip`. Ignoring `.src.zip` for the moment, `language` seems like something that the `Description` could figure out, so I don't see why this is a big deal. I would rather ban the `language` param quickly before people start relying on it.

Also, I don't see `language` documented on https://buckbuild.com/rule/android_library.html =\

Comment (bolinfest) [inline test/com/facebook/buck/android/testdata/android_project/kotlin/com/sample/lib/BUCK.fixture]: @zbsz So specifically, no, there should not be any backwards-compatibility issues. I believe we would pass those other types of files to `javac` and the compiler should reject them and fail.

**Comment by sdwilsh** [comment]
Are we going to make sure we have Kotlin installed in the travis/appveyor jobs so these tests can run?

**Comment by dsyang** [comment]
Re auto detecting the language. Sure I can remove `JvmLanguage`

Re documentation. Yea I need to write some stuff up for Kotlin. Maybe do a
blog post too when this lands and gets tested.

Re testing & installing: I think we can do something like what @mikekap did
for scala: add a .buckconfig flag that points to a build target for the
compiler. Then we can check in a version of the compiler and use that to
run tests. Since its just a .jar we should be able to run this without a
problem on existing infrastructure.

File: src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java (hunk 1 above)
Comment (Coneko) [inline src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java]: Still have to break this one per line.

File: src/com/facebook/buck/android/KotlinAndroidLibraryCompiler.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  @Override
+  public boolean trackClassUsage(JavacOptions javacOptions) {
+    return false;
+  }
+
+  @Override
+  public CompileToJarStepFactory compileToJar(
+      AndroidLibraryDescription.Arg args, JavacOptions javacOptions, BuildRuleResolver resolver) {
```
Comment (Coneko) [inline src/com/facebook/buck/android/KotlinAndroidLibraryCompiler.java]: One per line.

**Comment by Coneko** [comment]
Seems good, but split out the changes to remove the language and use the file extensions into a separate PR.

File: src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java [hunk 10]
```diff
@@ -16,23 +16,39 @@
...
+    this.kotlinBuckConfig = kotlinBuckConfig;
   }
 
   @Override
-  public AndroidLibraryCompiler getCompiler(AndroidLibraryDescription.JvmLanguage language) {
-    switch (language) {
-      case JAVA: return new JavaAndroidLibraryCompiler();
-      case SCALA: return new ScalaAndroidLibraryCompiler(scalaConfig);
+  public AndroidLibraryCompiler getCompiler(AndroidLibraryDescription.Arg args) {
+    // We currently expect all non-java languages to only contain files of that extension.
```
Comment (Coneko) [inline src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java]: Do we check this assumption anywhere? If not, we should, and throw a `HumanReadableException` that explains what is wrong.

**Comment by dsyang** [comment]
Removed changes for removing the language param. Will put out the new PR after this lands.

Something's broken with the LogFormatter which causes buck builds to fail. Not sure what though since this code doesn't touch that at all

**Pull request closed without merging**